gtfs/spec/en/reference.md

```diff
@@ -612,6 +624,69 @@ Describes levels in a station. Useful in conjunction with `pathways.txt`, and is
 |  `level_index` | Float | **Required** | Numeric index of the level that indicates its relative position. <br><br>Ground level should have index `0`, with levels above ground indicated by positive indices and levels below ground by negative indices.|
 |  `level_name` | Text | Optional | Name of the level as seen by the rider inside the building or station.<hr>_Example: Take the elevator to "Mezzanine" or "Platform" or "-1"._|
 
+### location_groups.txt
+
+File: **Optional**
+
+Primary key (`location_group_id`, `location_id`)
+
+Defines location groups that indicate groups of stops where a rider may request pickup or drop off.
+
+| Field Name | Type | Required | Description |
+| ---------- | ---- | ------------ | ----------- |
+| `location_group_id` | ID | **Required** | Identifies a location group. ID must be unique across all `stops.stop_id`, locations.geojson `id`, and `location_groups.location_group_id` values. <br><br>A location group is a group of stops that together indicate locations where a rider may request pickup or drop off.<br><br>Multiple entries in `location_groups.txt` can have the same `location_group_id`. | 
```

Comment:
Adding something probably relevant here, would love to hear other folks' thoughts.

In our last working group meeting, it was decided to move the table defining stop groups from being shared with fares v2 in `stop_areas.txt` back to using a dedicated file `location_groups.txt`. However, their mechanisms differ:

`stop_areas.txt` has a companion, `areas.txt`. `stop_areas` only assign stops to areas, and multiple entries in `stop_areas` can have the same `area_id`, without including any attributes. The `area_id` in `areas.txt` is a unique ID, and areas.txt contains attributes like `area_name`. When other tables reference the `area_id`, they reference `areas.area_id` rather than `stop_areas.area_id`.

However, the current `location_groups.txt` combines both "grouping" and "indicating attributes" within a single table.

What if we introduce a file similar to `areas.txt` for `location_groups.txt`? This could:
- Ensure a mechanism similar to existing `stop_areas`- `areas` and `route_networks` - `networks` in GTFS (even though they serve different purposes).
- Avoid duplicating values for attributes (such as location_group_name) within a single `location_groups` file.
- **Perhaps alleviate concerns raised in [this comment](https://github.com/google/transit/pull/388#discussion_r1416728745)? (Because the IDs in the new file would be unique ID.)**

Reply:
> Wording here could be confusing. This says the location_group_id must be unique "across all values" in three files, but then says multiple records in this file (rows in this table) may have the same location_group_id. To create a useful group, of course multiple rows in this table must reference the same location_group_id, which can be seen as the key for an implicit location_group entity that is not materialized in a table. The primary keys of those implicit location_groups then share a namespace with the primary keys of two other tables. The situation is unusual enough that wording will be important to avoid confusing people with more conventional data modeling expectations.

This wording was recently changed to be clearer than it was before. Suggestions to further improve the wording would definitely be welcome! The main point is we want to communicate here is that `stop_id`s, location.geojson `id`s, and `location_group_id`s cannot share the same value (e.g., a `stop_id` and location `id` cannot both equal `abc123`). In the case of `location_group_id`, multiple instances of a given `location_group_id` in location_groups.txt indicate that both those records are within a single location_group.

Regarding the rest:

From a producer perspective, it is more laborious to have to maintain single location groups using **two** files. I'll have to take your word that this feature deviates from conventional structuring of data, but Flex as a spec itself deviates from convention, to use @leonardehrenfried's words, these considerations are the easy part of implementing Flex. We really could spend an endless amount of time iterating components of the spec like this, but at the end of the day, consumers will still need to contend with the bigger challenge of implementing flexible routing. But I still wonder if the two file approach is less complex for a consumer and if mitigating that complexity outweighs the increased complexity for producers? To my knowledge, no Flex consumer since Flex v2's publication over three years ago has ever raised this as an inherent issue of location groups. Yes, areas/stop_areas.txt was momentarily used instead, but for the sole reason of reducing redundancy of file functionality, not because it used two files.

Reply:
A potential risk for producers could be that if this file includes more attributes in the future, it will require more repetitive values for these attributes. (Just raising a "potential" risk here, not strongly suggesting two separate files)

Also curious about how this is perceived from the consumer's standpoint.

Reply:
The spec as it is currently written addresses repetitive values with the following language for `location_group_name`: "Must be defined either once, or exhaustively for a single location_group_id." Is this a sufficient solution?

Reply:
Ah you are right, I didn't notice that. 
Yes then this won't be a risk for producers.

Reply:
> In our last working group meeting, it was decided to move the table defining stop groups from being shared with fares v2 in `stop_areas.txt` back to using a dedicated file `location_groups.txt`.

This could be justified just on the basis of them being different optional, self-contained spec extensions. Again I'm thinking about gtfs-components here. The way GTFS and its spec are broken into modules or components would need to be finalized before it would be clear if this is sufficient reason to have separate tables.

> `stop_areas.txt` has a companion, `areas.txt`. `stop_areas` only assign stops to areas, and multiple entries in `stop_areas` can have the same `area_id`, without including any attributes. The `area_id` in `areas.txt` is a unique ID, and areas.txt contains attributes like `area_name`. When other tables reference the `area_id`, they reference `areas.area_id` rather than `stop_areas.area_id`.

Something like `areas.txt` you are describing here is what I meant in my other comments by "materializing" the implicit location_groups in a table. This seems like a normalized relational model. As such it will be readily approachable and easier to grasp for many people. Arguably, it's not just "easier", but it only becomes possible to validate and reason about clearly when it adheres to certain principles. 

Just to be sure: the values appearing in `areas.area_id` and those in `stop_areas.area_id` are drawn from exactly the same set of values right? The set is just defined by the unique values in `areas`, then those same values are referenced in `stop_areas`? In standard terminology, they are the primary keys of `areas` and foreign keys in `stop_areas`? If so yes, this makes sense to me (though I find the naming of the tables a bit odd, something like `area_stops` seems more conventional).

This fare-oriented area-of-stops structure does seem very similar to the one that would be needed in flex routing though, so it is certainly worth asking if you need/want separate tables for the two. Strict componentization of GTFS seems like the main justification. If the attributes of areas/zones/groups are expected to remain identical for the fare and flex use cases, they can just share the tables. If the cases are (even potentially) divergent enough they can have different tables, at the cost of making the spec more complex but more modular.

> However, the current `location_groups.txt` combines both "grouping" and "indicating attributes" within a single table.

Right, it seems like a denormalized form of the fare-oriented area-of-stops.

> What if we introduce a file similar to `areas.txt` for `location_groups.txt`? 

This would not only be more consistent with other parts of GTFS (other "gtfs-components" you could say), it would be more consistent with decades of practice and theory underpinning databases and data modeling. Not necessarily introduce a file, but at least introduce the abstract entity (see below).

> * **Perhaps alleviate concerns raised in [this comment](https://github.com/google/transit/pull/388#discussion_r1416728745)? (Because the IDs in the new file would be unique ID.)**

Yes, I meant that the wording is probably confusing because the data model does not enable it to be clear. Changing some words in the description wouldn't change the structure they're trying to describe. Introducing distinct types/tables and normalizing the data (i.e. factoring redundant data or implicit entities out into their own tables) might allow for wording that is less confusing to the reader because it is able to refer more closely to the structure of the data in the files people see in front of them. On the other hand, if there aren't any redundant attributes to be factored out into this new table (in this case, the flex equivalent of fare-oriented areas) other than the IDs, you could elect to not materialize it. But in that case the entity *still exists in the model*, and should still be described in the documentation, it is just elided from the physical tables in the feed for convenience of feed producers (though It's debatable how convenient that is if they already have the table in their software). To be specific, you can say "There is this entity called a location_group, and each one has an ID that is unique among all location_groups, stops, and locations. However, these entities have no properties of their own, so they currently don't need a table. They're implicitly instantiated by reference from foreign keys in location_groups.txt." or to drop the jargon, "created when someone references their ID in this other table". 

Of course I still think renaming these things makes sense. The table location_groups would more rightly be the flex equivalent of your fare-oriented areas, and a table called something like location_group_stops would then associate stops to location_groups. In my opinion this all gets much less confusing if the tables are called flex_zones and flex_zone_stops. And even less confusing if you also have consistent naming with fare_areas and fare_area_stops.

I want to emphasize that I'm not making this stuff up as I go along or asking people to adopt "my way". I am just trying to describe standard data modeling and database design as practiced and refined for the last 50 years and taught in every computer science curriculum. It's founded in mathematics and decades of reasoning, not my personal opinion. Anything I say here may be flawed, and should be confronted with this standard practice and theory.

People with any background in data modeling or databases are already performing this mental transform into clear, standard language every time they read the spec. It's not so bad to convert the language, but it can get really frustrating when the underlying structure is too idiosyncratic or ad-hoc.

Reply:
"Materialized" implicit location_group in a separate file - `location_groups.txt` - https://github.com/google/transit/pull/388/commits/5e22e1062c2802ce591722e8c47559f011de3ca0

Resolving this conversation.